File: backend/database/mock_data.py
"""
Mock data provider for the ICS Security Monitoring System.
Provides sample data when database is not available.
"""

from datetime import datetime, timedelta
from typing import List, Dict, Any
import random


def get_mock_devices() -> List[Dict[str, Any]]:
    """Get mock device data."""
    return [
        {
            "id": 1,
            "ip_address": "192.168.95.2",
            "mac_address": "00:1A:2B:3C:4D:5E",
            "hostname": "plc_2",
            "device_type": "PLC",
            "vendor": "Siemens",
            "model": "S7-1200",
            "protocols": [{"id": 1, "name": "Modbus"}],
            "is_online": True,
            "risk_score": 85.0,
            "last_seen": datetime.utcnow(),
            "first_discovered": datetime.utcnow() - timedelta(days=30),
            "notes": "Primary PLC controlling production line"
        },
        {
            "id": 2,
            "ip_address": "192.168.95.3",
            "mac_address": "00:1A:2B:3C:4D:6F",
            "hostname": "hmi_station",
            "device_type": "HMI",
            "vendor": "Allen-Bradley",
            "model": "PanelView Plus",
            "protocols": [{"id": 2, "name": "EtherNet/IP"}],
            "is_online": True,
            "risk_score": 45.0,
            "last_seen": datetime.utcnow() - timedelta(minutes=2),
            "first_discovered": datetime.utcnow() - timedelta(days=25),
            "notes": "Operator interface station"
        },
        {
            "id": 3,
            "ip_address": "192.168.95.4",
            "mac_address": "00:1A:2B:3C:4D:70",
            "hostname": "scada_server",
            "device_type": "SCADA",
            "vendor": "Schneider Electric",
            "model": "Citect",
            "protocols": [{"id": 3, "name": "DNP3"}, {"id": 4, "name": "Modbus"}],
            "is_online": True,
            "risk_score": 65.0,
            "last_seen": datetime.utcnow() - timedelta(minutes=1),
            "first_discovered": datetime.utcnow() - timedelta(days=45),
            "notes": "Central SCADA system"
        }
    ]


def get_mock_alerts() -> List[Dict[str, Any]]:
    """Get mock alert data."""
    return [
        {
            "id": 1,
            "device_id": 1,
            "alert_type": "Buffer Overflow Attempt",
            "severity": "critical",
            "description": "Potential exploitation of libmodbus vulnerability on PLC_2",
            "details": {
                "source_ip": "192.168.95.100",
                "target_port": 502,
                "function_code": 16,
                "payload_size": 2048
            },
            "timestamp": datetime.utcnow() - timedelta(minutes=15),
            "acknowledged": False,
            "resolved": False
        },
        {
            "id": 2,
            "device_id": 2,
            "alert_type": "Unauthorized Access",
            "severity": "high",
            "description": "Failed login attempts detected on HMI station",
            "details": {
                "source_ip": "192.168.95.150",
                "failed_attempts": 5,
                "last_attempt": datetime.utcnow() - timedelta(hours=1)
            },
            "timestamp": datetime.utcnow() - timedelta(hours=2),
            "acknowledged": True,
            "acknowledged_by": "admin",
            "acknowledged_at": datetime.utcnow() - timedelta(hours=1),
            "resolved": False
        },
        {
            "id": 3,
            "device_id": 1,
            "alert_type": "Protocol Anomaly",
            "severity": "medium",
            "description": "Unusual Modbus function code sequence detected",
            "details": {
                "function_codes": [1, 3, 16, 23],
                "frequency": "high",
                "pattern": "suspicious"
            },
            "timestamp": datetime.utcnow() - timedelta(hours=6),
            "acknowledged": False,
            "resolved": False
        },
        {
            "id": 4,
            "device_id": 3,
            "alert_type": "Malicious Traffic",
            "severity": "high",
            "description": "Potential malware communication detected",
            "details": {
                "external_ip": "203.0.113.45",
                "port": 443,
                "data_size": "10MB",
                "encryption": "unknown"
            },
            "timestamp": datetime.utcnow() - timedelta(hours=12),
            "acknowledged": True,
            "acknowledged_by": "security_team",
            "acknowledged_at": datetime.utcnow() - timedelta(hours=10),
            "resolved": True,
            "resolved_at": datetime.utcnow() - timedelta(hours=8)
        }
    ]


def get_mock_connections() -> List[Dict[str, Any]]:
    """Get mock connection data."""
    return [
        {
            "id": 1,
            "source_id": 1,
            "target_id": 2,
            "protocol": "Modbus",
            "port": 502,
            "is_active": True,
            "traffic_volume": 150,
            "last_seen": datetime.utcnow(),
            "first_discovered": datetime.utcnow() - timedelta(days=20)
        },
        {
            "id": 2,
            "source_id": 2,
            "target_id": 1,
            "protocol": "EtherNet/IP",
            "port": 44818,
            "is_active": True,
            "traffic_volume": 89,
            "last_seen": datetime.utcnow() - timedelta(minutes=5),
            "first_discovered": datetime.utcnow() - timedelta(days=20)
        },
        {
            "id": 3,
            "source_id": 3,
            "target_id": 1,
            "protocol": "DNP3",
            "port": 20000,
            "is_active": True,
            "traffic_volume": 45,
            "last_seen": datetime.utcnow() - timedelta(minutes=2),
            "first_discovered": datetime.utcnow() - timedelta(days=30)
        }
    ]


def get_mock_protocol_stats() -> Dict[str, int]:
    """Get mock protocol statistics."""
    return {
        "Modbus": 450,
        "EtherNet/IP": 230,
        "DNP3": 150,
        "OPC-UA": 80,
        "BACnet": 40,
        "S7comm": 25
    }


def get_mock_traffic_volume(hours: int = 24) -> List[Dict[str, Any]]:
    """Get mock traffic volume data."""
    data_points = []
    now = datetime.utcnow()
    
    for i in range(hours, 0, -1):
        timestamp = now - timedelta(hours=i)
        
        # Generate realistic traffic patterns (higher during business hours)
        hour = timestamp.hour
        if 8 <= hour <= 18:  # Business hours
            base_traffic = 80
        elif 6 <= hour <= 8 or 18 <= hour <= 22:  # Transition hours
            base_traffic = 40
        else:  # Night hours
            base_traffic = 15
        
        # Add some randomness
        packet_count = base_traffic + random.randint(-20, 20)
        byte_count = packet_count * random.randint(50, 200)
        
        data_points.append({
            "id": i,
            "device_id": 1,
            "protocol": "Modbus",
            "packet_count": max(0, packet_count),
            "byte_count": max(0, byte_count),
            "timestamp": timestamp,
            "hour_bucket": timestamp.replace(minute=0, second=0, microsecond=0)
        })
    
    return data_points


def get_mock_detection_result(analysis_id: str) -> Dict[str, Any]:
    """Get mock detection result."""
    return {
        "id": 1,
        "analysis_id": analysis_id,
        "status": "completed",
        "start_time": datetime.utcnow() - timedelta(hours=1),
        "end_time": datetime.utcnow() - timedelta(minutes=50),
        "time_range": {
            "start": (datetime.utcnow() - timedelta(days=1)).isoformat(),
            "end": datetime.utcnow().isoformat()
        },
        "devices": [1, 2, 3],
        "anomalies": [
            {
                "device_id": 1,
                "timestamp": (datetime.utcnow() - timedelta(hours=3)).isoformat(),
                "anomaly_type": "traffic_spike",
                "score": 0.92,
                "description": "Unusual packet size detected in Modbus traffic"
            },
            {
                "device_id": 2,
                "timestamp": (datetime.utcnow() - timedelta(hours=2)).isoformat(),
                "anomaly_type": "protocol_violation",
                "score": 0.78,
                "description": "Invalid EtherNet/IP command sequence"
            }
        ],
        "summary": f"Analysis completed successfully. Found 2 anomalies across 3 devices."
    }

File: frontend/src/services/mockData.js
/**
 * Mock Data Service
 * Provides synthetic data for demonstration purposes
 */

// Generate a random IP address
const randomIP = () => {
  return `192.168.${Math.floor(Math.random() * 10)}.${Math.floor(Math.random() * 255)}`;
};

// Generate a random timestamp within the last 24 hours
const randomTimestamp = (hoursBack = 24) => {
  const date = new Date();
  date.setHours(date.getHours() - Math.random() * hoursBack);
  return date.toISOString();
};

// Device types common in ICS environments
const deviceTypes = ['plc', 'hmi', 'rtu', 'gateway', 'historian', 'workstation', 'scada_server'];

// Common industrial protocols
const protocols = [
  { name: 'Modbus TCP', id: 1 },
  { name: 'EtherNet/IP', id: 2 },
  { name: 'Profinet', id: 3 },
  { name: 'DNP3', id: 4 },
  { name: 'OPC UA', id: 5 },
  { name: 'BACnet', id: 6 },
  { name: 'S7', id: 7 }
];

// Alert types
const alertTypes = [
  'Unauthorized Access',
  'Unusual Traffic Pattern',
  'Protocol Violation',
  'Device Configuration Change',
  'Firmware Modification',
  'Command Injection',
  'DDoS Attack',
  'Port Scan',
  'Malformed Packet'
];

// Alert severity levels
const severityLevels = ['low', 'medium', 'high', 'critical'];

// Create mock devices
const generateDevices = (count = 15) => {
  const devices = [];
  
  for (let i = 0; i < count; i++) {
    const deviceType = deviceTypes[Math.floor(Math.random() * deviceTypes.length)];
    const isOnline = Math.random() > 0.2;
    const deviceProtocols = [];
    const protocolCount = Math.floor(Math.random() * 3) + 1;
    
    // Assign random protocols to this device
    const shuffledProtocols = [...protocols].sort(() => 0.5 - Math.random());
    for (let j = 0; j < protocolCount; j++) {
      if (j < shuffledProtocols.length) {
        deviceProtocols.push(shuffledProtocols[j]);
      }
    }
    
    devices.push({
      id: i + 1,
      ip_address: randomIP(),
      hostname: `${deviceType}-${i + 1}`,
      device_type: deviceType,
      protocols: deviceProtocols,
      is_online: isOnline,
      risk_score: Math.random() * 100,
      last_seen: randomTimestamp(isOnline ? 1 : 48)
    });
  }
  
  return devices;
};

// Create mock alerts
const generateAlerts = (devices, count = 20) => {
  const alerts = [];
  
  for (let i = 0; i < count; i++) {
    const severity = severityLevels[Math.floor(Math.random() * severityLevels.length)];
    const device = devices[Math.floor(Math.random() * devices.length)];
    const alertType = alertTypes[Math.floor(Math.random() * alertTypes.length)];
    const acknowledged = Math.random() > 0.7;
    
    alerts.push({
      id: i + 1,
      device_id: device.id,
      device_ip: device.ip_address,
      alert_type: alertType,
      severity: severity,
      description: `${alertType} detected on ${device.device_type} (${device.ip_address})`,
      timestamp: randomTimestamp(24),
      acknowledged: acknowledged,
      acknowledged_by: acknowledged ? 'admin' : null
    });
  }
  
  // Sort by timestamp (newest first)
  return alerts.sort((a, b) => new Date(b.timestamp) - new Date(a.timestamp));
};

// Create mock network connections
const generateNetworkMap = (devices) => {
  const connections = [];
  
  // Create a central switch/router
  const centralNode = {
    id: 'central-switch',
    ip_address: '192.168.1.1',
    hostname: 'central-switch',
    device_type: 'network',
    is_online: true,
    risk_score: 10
  };
  
  // Add central node to devices
  const allDevices = [...devices, centralNode];
  
  // Connect all devices to central node
  devices.forEach(device => {
    connections.push({
      source: 'central-switch',
      target: device.id,
      traffic_volume: Math.floor(Math.random() * 100) + 10
    });
    
    // Add some device-to-device connections
    if (Math.random() > 0.7) {
      const otherDevice = devices.find(d => d.id !== device.id && Math.random() > 0.5);
      if (otherDevice) {
        connections.push({
          source: device.id,
          target: otherDevice.id,
          traffic_volume: Math.floor(Math.random() * 50) + 5
        });
      }
    }
  });
  
  return { devices: allDevices, connections };
};

// Generate traffic data for the past 24 hours
const generateTrafficData = (hourCount = 24, pointsPerHour = 4) => {
  const data = [];
  const now = new Date();
  
  for (let hour = hourCount; hour >= 0; hour--) {
    for (let point = 0; point < pointsPerHour; point++) {
      const timestamp = new Date(now);
      timestamp.setHours(now.getHours() - hour);
      timestamp.setMinutes(Math.floor((point / pointsPerHour) * 60));
      
      // Create some patterns in the data
      const timeOfDay = timestamp.getHours();
      let multiplier = 1;
      
      // More traffic during work hours
      if (timeOfDay >= 8 && timeOfDay <= 17) {
        multiplier = 2.5;
      }
      
      // Add some randomness
      const randomFactor = 0.7 + Math.random() * 0.6;
      
      const packetCount = Math.floor(
        500 * multiplier * randomFactor + (Math.sin(timeOfDay / 3) * 200)
      );
      
      const byteCount = packetCount * (64 + Math.floor(Math.random() * 1000));
      
      data.push({
        timestamp: timestamp.toISOString(),
        packet_count: packetCount,
        byte_count: byteCount,
      });
    }
  }
  
  return data;
};

// Generate protocol statistics
const generateProtocolStats = () => {
  const stats = {};
  
  protocols.forEach(protocol => {
    stats[protocol.name] = Math.floor(Math.random() * 1000) + 100;
  });
  
  return stats;
};

// Main export
const mockData = {
  devices: generateDevices(15),
  get alerts() {
    return generateAlerts(this.devices, 20);
  },
  get networkMap() {
    return generateNetworkMap(this.devices);
  },
  trafficData: generateTrafficData(),
  protocolStats: generateProtocolStats()
};

export default mockData; 